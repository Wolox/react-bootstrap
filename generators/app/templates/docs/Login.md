## Login

**Props**

| Property | Type | Description |
|:---|:---:|:---:|
| handleEmailChange | Function | Function to be called on email change |
| handlePasswordChange | Function | Function to be called on password change |
| handleLogin | Function | Function to be called on form submit |

**Usage**

```jsx
  <Login
    handleEmailChange={this.handleEmailChange}
    handlePasswordChange={this.handlePasswordChange}
    handleLogin={this.handleLogin}
  />
````
