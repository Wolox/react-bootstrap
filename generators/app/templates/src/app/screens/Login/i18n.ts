import i18next from 'i18next';

i18next.addResources('es', 'Login', {
  login: 'Iniciar sesión',
  loginExplanation: 'Ingrese los siguientes datos',
  enter: 'Ingresar',
  forgotPassword: 'Olvidé mi contraseña',
  email: 'E-mail',
  emailPlaceholder: 'Ej: nombre@empresa.com',
  password: 'Contraseña',
  passwordPlaceholder: 'Ingrese su contraseña',
  createAccount: 'Crear cuenta',
  invalidCredentials: 'Credenciales incorrectas',
  error: 'Ocurrió un error en el login'
});
