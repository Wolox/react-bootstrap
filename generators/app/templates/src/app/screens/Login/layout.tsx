import React from 'react';
import i18next from 'i18next';

import FormInput from '~components/FormInput';
import PATHS from '~components/Routes/paths';
import { withSpinner } from '~components/Spinner';
import { LoginError } from '~services/AuthServices';
import { Nullable } from '~utils/types';
import { Error } from '~app/hooks/useRequest';
import { stringArrayToObject } from '~utils/array';

import styles from './styles.module.scss';

interface Props {
  onEmailChange: (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>) => void;
  onLogin: (event: React.FormEvent<HTMLFormElement>) => void;
  onPasswordChange: (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>) => void;
  loading?: boolean;
  loginError?: Nullable<Error<LoginError>>;
}

const FIELDS = stringArrayToObject(['email', 'password']);
function Login({ onEmailChange, onPasswordChange, onLogin, loginError, loading }: Props) {
  const errorMessage = loginError?.errorData?.message;
  return (
    <div className={`column center full-width ${styles.container}`}>
      <div className="column center m-bottom-3">
        <h1 className="m-bottom-1">{i18next.t('Login:login')}</h1>
        <h2>{i18next.t('Login:loginExplanation')}</h2>
      </div>
      <form className={`column m-bottom-2 ${styles.formContainer}`} onSubmit={onLogin}>
        <FormInput
          label={i18next.t('Login:email')}
          name={FIELDS.email}
          inputType="text"
          inputClassName={`m-bottom-2 full-width ${styles.input}`}
          placeholder={i18next.t('Login:emailPlaceholder') as string}
          onChange={onEmailChange}
          disabled={loading}
        />
        <FormInput
          label={i18next.t('Login:password')}
          name={FIELDS.password}
          inputType="password"
          inputClassName={`m-bottom-2 full-width ${styles.input}`}
          placeholder={i18next.t('Login:passwordPlaceholder') as string}
          onChange={onPasswordChange}
          disabled={loading}
        />
        <div className="column center">
          <button disabled={loading} type="submit" className={`full-width m-bottom-1 ${styles.button}`}>
            {i18next.t('Login:enter')}
          </button>
          <span className={`row center middle full-width m-top-1 m-bottom-1 ${errorMessage ? '' : 'hidden'}`}>
            {i18next.t(`Login:${errorMessage || 'error'}`)}
          </span>
          <a href={PATHS.recoverPassword}>{i18next.t('Login:forgotPassword')}</a>
          <a href={PATHS.registration}>{i18next.t('Login:createAccount')}</a>
        </div>
      </form>
    </div>
  );
}

export default withSpinner({ classNameContainer: styles.loading })(Login);
