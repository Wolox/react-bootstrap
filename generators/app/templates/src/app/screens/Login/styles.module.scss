@import 'variables/colors';

.button {
  border: solid 1px;
  border-radius: 5px;
  height: 40px;
}

.input {
  border: solid 1px;
  border-radius: 5px;
  height: 30px;
  padding: 5px;
}

.container {
  min-height: 500px;
  padding-top: 80px;
}

.form-container {
  max-width: 280px;
  width: 100%;
}

.error-message {
  background-color: $red;
  border-radius: 4px;
  color: $white;
  height: 50px;
  font-size: 20px;
  font-weight: bold;
}

.loading {
  height: 100px;
  left: calc(50% - 50px);
  position: absolute;
  top: 50%;
  width: 100px;
}
