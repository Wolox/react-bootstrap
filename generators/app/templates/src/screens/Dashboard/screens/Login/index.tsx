function Login() {
  
  return (
    <div>
      <span>login</span>
    </div>
  );
}

export default Login;
