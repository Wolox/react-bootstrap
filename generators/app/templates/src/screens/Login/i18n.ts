import i18next from 'i18next';

i18next.addResources('es', 'Login', {
  login: 'Iniciar sesión',
  loginExplanation: 'Ingrese los siguientes datos',
  enter: 'Ingresar',
  forgotPassword: 'Olvidé mi contraseña',
  email: 'E-mail',
  emailPlaceholder: 'Ej: nombre@empresa.com',
  password: 'Contraseña',
  passwordPlaceholder: 'Ingrese su contraseña',
  createAccount: 'Crear cuenta',
  invalidCredentials: 'Credenciales incorrectas',
  error: 'Ocurrió un error en el login',
  emailFormatError: 'El formato es inválido',
  requiredError: 'El campo es requerido',
  changeLang: 'Cambiar de idioma'
});

i18next.addResources('en', 'Login', {
  login: 'Login',
  loginExplanation: 'Enter the following information',
  enter: 'Enter',
  forgotPassword: 'Forgot password',
  email: 'E-mail',
  emailPlaceholder: 'Eg: name@company.com',
  password: 'Password',
  passwordPlaceholder: 'Enter password',
  createAccount: 'Create account',
  invalidCredentials: 'Incorrect credentials',
  error: 'There was an error on login',
  emailFormatError: 'Invalid format',
  requiredError: 'Field is required',
  changeLang: 'Change language'
});
