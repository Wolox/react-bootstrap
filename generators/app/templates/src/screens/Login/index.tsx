import React, { useCallback } from 'react';
import { useHistory } from 'react-router-dom';
import i18next from 'i18next';
import { useForm } from 'react-hook-form';

import { useLazyRequest } from 'hooks/useRequest';
import { useDispatch } from 'contexts/UserContext';
import { actionCreators, Credentials, User } from 'contexts/UserContext/reducer';
import { login, setCurrentUser } from 'services/AuthServices';
import FormInput from 'components/FormInput';
import PATHS from 'components/Routes/paths';

import styles from './styles.module.scss';

const FIELDS = { email: 'email', password: 'password' };

const VALIDATIONS = {
  email: {
    pattern: {
      value: /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
      message: i18next.t('Login:emailFormatError') as string
    },
    required: i18next.t('Login:requiredError') as string
  },
  password: { required: i18next.t('Login:requiredError') as string }
};

function LoginContainer() {
  const history = useHistory();
  const dispatch = useDispatch();

  const [, loading, loginError, loginRequest] = useLazyRequest({
    request: (credentials: Credentials) => login(credentials),
    withPostSuccess: response => {
      const userResponse = response as User;
      dispatch(actionCreators.setUser(userResponse));
      setCurrentUser(userResponse);

      history.push('/');
    }
  });
  const { register, handleSubmit, errors } = useForm();
  const errorMessage = loginError?.errorData?.message;

  const handleLogin = useCallback(
    (values: Credentials) => {
      loginRequest(values);
    },
    [loginRequest]
  );

  return (
    <div className={`column center full-width ${styles.container}`}>
      <div className="column center m-bottom-3">
        <h1 className="m-bottom-1">{i18next.t('Login:login')}</h1>
        <h2>{i18next.t('Login:loginExplanation')}</h2>
      </div>
      <form
        className={`column m-bottom-2 ${styles.formContainer}`}
        aria-label="login-form"
        onSubmit={handleSubmit(handleLogin)}
        noValidate
      >
        <FormInput
          label={i18next.t('Login:email')}
          name={FIELDS.email}
          inputType="email"
          inputClassName={`m-bottom-2 full-width ${styles.input}`}
          placeholder={i18next.t('Login:emailPlaceholder') as string}
          disabled={loading}
          inputRef={register(VALIDATIONS.email)}
          error={errors?.email?.message}
        />
        <FormInput
          label={i18next.t('Login:password')}
          name={FIELDS.password}
          inputType="password"
          inputClassName={`m-bottom-2 full-width ${styles.input}`}
          placeholder={i18next.t('Login:passwordPlaceholder') as string}
          disabled={loading}
          inputRef={register(VALIDATIONS.password)}
          error={errors?.password?.message}
        />
        <div className="column center">
          <button disabled={loading} type="submit" className={`full-width m-bottom-1 ${styles.button}`}>
            {i18next.t('Login:enter')}
          </button>
          <span className={`row center middle full-width m-top-1 m-bottom-1 ${errorMessage ? '' : 'hidden'}`}>
            {i18next.t(`Login:${errorMessage || 'error'}`)}
          </span>
          <a href={PATHS.recoverPassword}>{i18next.t('Login:forgotPassword')}</a>
          <a href={PATHS.registration}>{i18next.t('Login:createAccount')}</a>
        </div>
      </form>
    </div>
  );
}

export default LoginContainer;
