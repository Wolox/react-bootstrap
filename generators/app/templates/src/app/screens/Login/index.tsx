import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';

import { useDispatch } from '~contexts/UserContext';
import { actionCreators, Credentials, User } from '~contexts/UserContext/reducer';
import { useLazyRequest } from '~app/hooks/useRequest';
import { login, setCurrentUser } from '~services/AuthServices';

import Login from './layout';

function LoginContainer() {
  const history = useHistory();
  const dispatch = useDispatch();
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [, loading, loginError, loginRequest] = useLazyRequest({
    request: (credentials: Credentials) => login(credentials),
    withPostSuccess: response => {
      const userResponse = response as User;
      dispatch(actionCreators.setUser(userResponse));
      setCurrentUser(userResponse);
      history.push('/');
    }
  });

  const handleLogin = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const credentials = { username: email, password };
    dispatch(actionCreators.login(credentials));
    loginRequest(credentials);
  };

  const handleEmailChange = (e: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>) =>
    setEmail((e.target as HTMLInputElement).value || '');

  const handlePasswordChange = (e: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    setPassword((e.target as HTMLInputElement).value || '');
  };

  return (
    <Login
      onEmailChange={handleEmailChange}
      onPasswordChange={handlePasswordChange}
      onLogin={handleLogin}
      loading={loading}
      loginError={loginError}
    />
  );
}

export default LoginContainer;
