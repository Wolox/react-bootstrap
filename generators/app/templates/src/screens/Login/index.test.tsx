import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import 'mutationobserver-shim';

import Login from './index';

const mockSetStateUser = jest.fn();
const mockSetPersistantUser = jest.fn();

jest.mock('~contexts/UserContext/reducer', () => ({
  actionCreators: {
    setUser: (values: any) => mockSetStateUser(values)
  }
}));

jest.mock('~services/AuthService', () => ({
  login: () =>
    new Promise(resolve =>
      resolve({
        ok: true,
        data: { sessionToken: 'token', id: 1234 },
        problem: null,
        originalError: null
      })
    ),
  setCurrentUserToken: (values: any) => mockSetPersistantUser(values)
}));

describe('#Login', () => {
  const component = (
    <Router history={createMemoryHistory()}>
      <Login />
    </Router>
  );

  describe('when mounting', () => {
    it('shows valid content', () => {
      const { container } = render(component);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('when filling invalid fields without submitting', () => {
    it('does not show the field errors', async () => {
      const { container, getByLabelText } = render(component);
      const email = getByLabelText('Login:email');

      await fireEvent.change(email, { target: { value: 'invalid email' } });

      expect(container.innerHTML).not.toMatch('Login:emailFormatError');
    });
  });

  describe('when filling invalid email and no password and submitting', () => {
    it('shows the field errors', async () => {
      const { container, getByRole, getByLabelText } = render(component);
      const email = getByLabelText('Login:email');
      const form = getByRole('form', { name: 'login-form' });

      await fireEvent.change(email, { target: { value: 'invalid email' } });

      // eslint-disable-next-line max-nested-callbacks
      await waitFor(() => fireEvent.submit(form));

      expect(container.innerHTML).toMatch('Login:emailFormatError');
      expect(container.innerHTML).toMatch('Login:required');
    });
  });

  describe('when empty email and no password and submitting', () => {
    it('shows the field errors', async () => {
      const { container, getByRole } = render(component);
      const form = getByRole('form', { name: 'login-form' });

      // eslint-disable-next-line max-nested-callbacks
      await waitFor(() => fireEvent.submit(form));

      expect(container.innerHTML).toMatch('Login:required');
    });
  });

  describe('when valid email and password', () => {
    it('executes the request and saves user', async () => {
      const { getByRole, getByLabelText } = render(component);
      const email = getByLabelText('Login:email');
      const password = getByLabelText('Login:password');
      const form = getByRole('form', { name: 'login-form' });
      await fireEvent.change(email, { target: { value: 'someone@wolox.com' } });
      await fireEvent.change(password, { target: { value: 'myPassword' } });

      // eslint-disable-next-line max-nested-callbacks
      await waitFor(() => fireEvent.submit(form));

      expect(mockSetStateUser).toHaveBeenCalledWith({ sessionToken: 'token', id: 1234 });
      expect(mockSetPersistantUser).toHaveBeenCalledWith({ sessionToken: 'token', id: 1234 });
    });
  });
});
