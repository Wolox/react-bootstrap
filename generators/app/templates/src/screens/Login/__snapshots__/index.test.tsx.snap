// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Login when mounting shows valid content 1`] = `
<div
  class="column center full-width container"
>
  <div
    class="column center m-bottom-3"
  >
    <h1
      class="m-bottom-1"
    >
      Login:login
    </h1>
    <h2>
      Login:loginExplanation
    </h2>
  </div>
  <form
    aria-label="login-form"
    class="column m-bottom-2 formContainer"
    novalidate=""
  >
    <div
      class="column start "
    >
      <label
        class=" m-bottom-1"
        for="email"
      >
        Login:email
      </label>
      <input
        class="m-bottom-2 full-width input input "
        id="email"
        name="email"
        placeholder="Login:emailPlaceholder"
        type="email"
      />
      <span
        aria-hidden="true"
        class=" errorText "
        id="email-error"
        role="alert"
      />
    </div>
    <div
      class="column start "
    >
      <label
        class=" m-bottom-1"
        for="password"
      >
        Login:password
      </label>
      <input
        class="m-bottom-2 full-width input input "
        id="password"
        name="password"
        placeholder="Login:passwordPlaceholder"
        type="password"
      />
      <span
        aria-hidden="true"
        class=" errorText "
        id="password-error"
        role="alert"
      />
    </div>
    <div
      class="column center"
    >
      <button
        class="full-width m-bottom-1 button"
        type="submit"
      >
        Login:enter
      </button>
      <span
        class="row center middle full-width m-top-1 m-bottom-1 hidden"
      >
        Login:error
      </span>
      <a
        href="/recover_password"
      >
        Login:forgotPassword
      </a>
      <a
        href="/registration"
      >
        Login:createAccount
      </a>
    </div>
  </form>
</div>
`;
